lib/vector-common/src/request_metadata.rs

```diff
@@ -94,5 +264,96 @@ impl<'a> Add<&'a RequestMetadata> for RequestMetadata {
 /// Objects implementing this trait have metadata that describes the request.
 pub trait MetaDescriptive {
     /// Returns the `RequestMetadata` associated with this object.
-    fn get_metadata(&self) -> RequestMetadata;
+    fn get_metadata(&self) -> &RequestMetadata;
+
+    /// Returns the owned `RequestMetadata`.
+    /// It is *very* important that this function is not called more than once for each
+    /// instance since it is likely that by taking the metadata it will replace the
+    /// current values with default metadata.
+    fn take_metadata(&mut self) -> RequestMetadata;
```

Comment:
For future consideration: If there is ever more that needs to be done to the metadata than just `take`, this should probably become `fn metadata_mut(&mut self) -> &mut RequestMetadata` and then move the `take` to the call site.